### DISCUSSION TITLE
"stabilizer loop rate is off" / "Kalman prediction rate low" Issue

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#639] [Q&A] [answered]
Hello.
I'm using the Crazyflie 2.1 and Flow Deck V2.
While experimenting while modifying the firmware code, I saw the following warning.
![image](https://user-images.githubusercontent.com/110527007/229066366-78a2a7d1-462b-447d-aaef-b80d57f20852.png)
In the previous discussion, if the rate is too far away from the expected 1000Hz, the warning will appear.
I have now made some modifications to attide_pid_controller.c and controller_pid.c.
The modified part is the Rate Controller part, and the modified part is as follows.
![image](https://user-images.githubusercontent.com/110527007/229067625-711a0daf-70c1-4457-a9b7-8b834231b47d.png)
I wonder if my approach is wrong.
The intent of the code is that if the rate value is greater than |240|deg/s, the PID controller of another gain is executed, or the controller of the existing firmware is executed.
Will this cause the warning in the picture above?

Ah, The reason I'm trying this is to recover the attitude when the crazyflie is spinning.

**Comment by krichardsson** [comment, marked as answer]
The warnings are displayed when the stabilizer task and the kalman estimator tasks do not get enough CPU-time to perform their job in time. The reason is related to that some code somewhere in your system is using too much CPU resources, but the code you have supplied looks OK to me and I don't think it is related to the problem. You might have some code somewhere else with a spin loop or heavy computations?

I'm not sure if this is a misunderstanding or coincidence, but note that the "rate" in the warnings refer to the rate that the stabilizer loop and estimator loop code runs at, that is the number of times per second that the code executes, while the "rate" in the code above (for instance `att_rate_control()`) is related to the change in attitude angles. They are not the same thing. If this was already clear, please disregard this comment, I just want to make sure :-)

**Reply by GNC-JY** [reply]
att_rate_control has the same configuration as the existing attitudeControllerCorrectRatePID function. Only Controller Gain values are different.
![image](https://user-images.githubusercontent.com/110527007/229085514-fd007325-1bfe-4c2e-a59f-08b5aa26b8fb.png)
I just want to use gyro to control the rotation when the crazyflie is spinning.
Can using the variable in stabilizer.c by external float in another .c file affect the rate? In other .c files, it is used as constant.

So... One more question, Does Flow Deck originally have a low prediction rate? When checked by several Crazyflies, the rate is 10 to 40.

**Reply by krichardsson** [reply]
> Can using the variable in stabilizer.c by external float in another .c file affect the rate? In other .c files, it is used as constant.

No, that should not be a problem

A good way to approach this type of problem is usually to comment out parts of your code to figure out what is causing the problem. Or start from a stable, working code base and add your stuff in, bit by bit, to find when you start to see the warnings.

**Reply by krichardsson** [reply]
> So... One more question, Does Flow Deck originally have a low prediction rate? When checked by several Crazyflies, the rate is 10 to 40.

Please start a new discussion for this question, as it does not fit with the title
